### PR TITLE
Implement RFC 69

### PR DESCRIPTION
**Pull request by whitequark**

### CODE REVIEW

File: amaranth/lib/io.py [hunk 1]
```diff
@@ -331,6 +314,167 @@ def __repr__(self):
...
+    @property
+    def invert(self):
+        return self._invert
+
+    @property
+    def direction(self):
+        return self._direction
+
+    def __len__(self):
+        if self._direction is Direction.Input:
```
Comment (wanda-phi) [inline amaranth/lib/io.py:417]: could just do `return len(self._invert)`

Comment (whitequark) [reply amaranth/lib/io.py:417]: That's... kind of cursed, though true.

File: amaranth/lib/io.py [hunk 2]
```diff
@@ -476,6 +620,16 @@ def elaborate(self, platform):
             else:
                 m.submodules += IOBufferInstance(self._port.p, o=o_inv, oe=self.oe, i=i_inv)
                 m.submodules += IOBufferInstance(self._port.n, o=~o_inv, oe=self.oe)
+        elif isinstance(self._port, SimulationPort):
+            if self.direction in (Direction.Input, Direction.Bidir):
+                # Loop back `o` if `oe` is asserted. This frees the test harness from having to
+                # provide this functionality itself.
+                for i_inv_bit, oe_bit, o_bit, i_bit in \
+                        zip(i_inv, self._port.oe, self._port.o, self._port.i):
```
Comment (wanda-phi) [inline amaranth/lib/io.py:628]: this would cause an `AttributeError` on ports with `Input` direction, wouldn't it?

Comment (whitequark) [reply amaranth/lib/io.py:628]: Oh, hm. I wonder why are the tests not catching that?

Comment (whitequark) [reply amaranth/lib/io.py:628]: Fixed

File: docs/changes.rst [hunk 3]
```diff
@@ -23,6 +23,14 @@ Version 0.6 (unreleased)
 ========================
 
 
+Implemented RFCs
```
Comment (wanda-phi) [inline docs/changes.rst:26]: given that we intend to backport it, shouldn't this be in 0.5.1 section?

Comment (whitequark) [reply docs/changes.rst]: moved

File: docs/changes.rst [hunk 4]
```diff
@@ -59,6 +67,8 @@ Standard library changes
 * Added: views of :class:`amaranth.lib.data.ArrayLayout` can be indexed with negative integers or slices.
 * Added: :py:`len()` works on views of :class:`amaranth.lib.data.ArrayLayout`.
 * Added: views of :class:`amaranth.lib.data.ArrayLayout` are iterable.
+* Added: :class:`io.SimulationPort`. (`RFC 69`_)
+* Changed: providing :meth:`io.PortLike.__add__` is now mandatory. (`RFC 69`_)
```
Comment (wanda-phi) [inline docs/changes.rst:71]: ... and that should actually be in 0.6?

File: docs/changes.rst [hunk 5]
```diff
@@ -44,13 +44,22 @@ Standard library changes
 .. currentmodule:: amaranth.lib
 
 * Added: :py:`payload_init=` argument in :class:`amaranth.lib.stream.Signature`.
+* Changed: (deprecated in 0.5) providing :meth:`io.PortLike.__add__` is now mandatory. (`RFC 69`_)
```
Comment (wanda-phi) [inline docs/changes.rst:47]:
```suggestion
* Changed: (deprecated in 0.5.1) providing :meth:`io.PortLike.__add__` is now mandatory. (`RFC 69`_)
```

Comment (whitequark) [reply docs/changes.rst:47]: I've thought about this but then we should probably also edit the other version references to say .0 at the end...

Comment (whitequark) [reply docs/changes.rst]: Done that.

File: tests/utils.py [hunk 6]
```diff
@@ -16,15 +16,57 @@
...
+                        output_repr.append(prefix)
+                    else:
+                        output_repr.append(char)
+                elif name is None:
+                    current += char
+                    output_repr.append(char)
+                else:
+                    output_repr.append(char)
+            return "".join(output_repr)
+        # print("\n" + format_repr(squish_repr(repr(obj))))
```
Comment (wanda-phi) [inline tests/utils.py:68]: leftover?

File: tests/utils.py [hunk 7]
```diff
@@ -16,15 +16,57 @@
...
         if isinstance(obj, list):
             obj = Statement.cast(obj)
-        def prepare_repr(repr_str):
+        def squish_repr(repr_str):
             repr_str = re.sub(r"\s+",   " ",  repr_str)
             repr_str = re.sub(r"\( (?=\()", "(", repr_str)
             repr_str = re.sub(r"\) (?=\))", ")", repr_str)
             return repr_str.strip()
-        self.assertEqual(prepare_repr(repr(obj)), prepare_repr(repr_str))
+        def format_repr(input_repr, *, indent="    "):
```
Comment (wanda-phi) [inline tests/utils.py:29]: horrifying.

File: tests/utils.py (hunk 6 above)
Comment (whitequark) [reply tests/utils.py:68]: No, intentionally left in there for the next person who needs to update NIR tests. You can look at the diff but the diff has annoying `- ` at the front of each line.

File: tests/utils.py (hunk 7 above)
Comment (whitequark) [reply tests/utils.py:29]: thank you

File: tests/utils.py (hunk 6 above)
Comment (wanda-phi) [reply tests/utils.py:68]: fair enough

**Review by wanda-phi — APPROVED** [review]

**Pull request merged**